enhancements/kubernetes-nmstate-integration.md

```diff
@@ -0,0 +1,165 @@
+with other names will be ignored.
+
+A new kubernetes-nmstate handler DaemonSet is deployed in the cluster by the operator.
+This DaemonSet contains nmstate package and interacts with the NetworkManager
+on the host by mounting the related dbus. The project contains three
+Custom Resource Definitions, `NodeNetworkState`, `NodeNetworkConfigurationPolicy` and
+`NodeNetworkConfigurationEnactment`. `NodeNetworkState` objects are created per each
+node in the cluster and can be used to report available interfaces and network configuration.
+These objects are created by kubernetes-nmstate and must not be touched by a user.
+`NodeNetworkConfigurationPolicy` objects can be used to specify desired
+networking state per node or set of nodes. It uses API similar to `NodeNetworkState`. 
+`NodeNetworkConfigurationEnactment` objects will be created per each node, per each matching
+`NodeNetworkConfigurationPolicy`.
+
+kubernetes-nmstate DaemonSet creates a custom resource of `NodeNetworkState` type per each node and
+updates the network topology from each OpenShift node.
+
+User configures host network changes and apply a policy in `NodeNetworkConfigurationPolicy` custom
+resource. Network topology is configured via `desiredState` section in `NodeNetworkConfigurationPolicy`.
+Multiple `NodeNetworkConfigurationPolicy` custom resources can be created.
+
+Upon receiving a notification event of `NodeNetworkState` update,
+kubernetes-nmstate Daemon verify the correctness of `NodeNetworkState` custom resource and
+apply the selected profile to the specific node.
+
+`NodeNetworkConfigurationEnactment` objects is a read-only object that represents Policy exectuion
+per each matching Node. It will expose configuration status per each Node.
+
+A new container image (kubernetes-nmstate-handler) will be created.
+
+A new container image (kubernetes-nmstate-operator) will be created. However, the operator
+and the handler will co-exist in the same upstream repo.
+
+An upstream API group of nmstate.io is currently used.
+
+### User Stories
+
+#### Bond creation
+
+* Be able to create bond interfaces on OpenShift nodes.
+ 
+  Typically, MCO + Ignition would be utilized. However, in a Day 1 install, that may not be possible.
+  It may be necessary for the bond/vlan configuration be done before Ignition can even be reached.
+  
+* Create a vlan interface on top of the bond inter.
+
+#### Assign ip address
+
+* Assign static and/or dynamic ip address on interfaces
+* Assign ipv4 and/or ipv6
+
+#### Create/Update/Remove network routes
+
+* Be able to Create/Update/Remove network routes for different interfaces like (bond,ethernet,sriov vf and sriov pf)
+
+#### Manage/Configure host SR-IOV network device
+
+* Be able to change host Virtual functions configuration (not managed by the sriov-operator) like vlans mtu driver etc..
+
+#### Rollback
+
+* Be able to rollback network configuration 
```

Comment:
This would be part of a meta-user-story that applies to all of them. Something like "I want to be able to roll out changes in a staged, safe manner without reboots. I am concerned that nodes will go unreachable."

Reply:
This is strongly related to the "integrate with MCO" path - if nmstate does things like roll out network configuration to multiple nodes at once that makes it wayyy too easy to take down a cluster.  Plus even if it doesn't, and e.g the MCO is starting to drain+reboot one master node while nmstate applies a broken networking change to a *different* master that's also very problematic.

Reply:
kubernetes-nmstate has an automatic rollback. If it detects that API is no longer reachable, it will rollback to a good state. This is already implemented.

Reply:
@bcrochet where can we go to learn more about this rollback?  how do we know it will not conflict with mco itself?

Reply:
Automatic rollbacks get tricky.  Have you thought about scenarios where e.g. a new network configuration lands, and then something *else* causes the API server to become temporarily unreachable, and then the problem becomes compounded by a rollback to a previous network configuration?

Reply:
@derekwaynecarr we haven't document rollback yet, usually the u/s documentation is at https://github.com/nmstate/kubernetes-nmstate, I will add a PR there to include the info related to rollback.

About rollback details kubernetes-nmstate try to apply some configuration with some timeout so if knmstate dies the networkmanager at nodes will rollback it, then it does a pair of "probes" one is related to ping external world (we do that by pinging default gw) and the other is realted to checking API server works fine (access to a mandatory namespace is done) if nothing related to this works for some time rollback to old config is done.

@cgwalters since something "else" has break API server you are as bad with old or new network config I suppose, even being with old at least you are at the starting point, but I am not sure, this is going to be quite different depending on what was being changed at network and what happend to API server.